components/locale_canonicalizer/src/locale_canonicalizer.rs

```diff
@@ -47,18 +120,276 @@ macro_rules! maximize_locale {
+                                    &rule.1[0]
+                                }
+                            } else {
+                                &rule.1[0]
+                            }
+                        } else {
+                            &rule.1[0]
+                        };
+                    if let Ok(replacement) = replacement.parse::<subtags::Region>() {
+                        locale.id.region = Some(replacement);
+                        result = CanonicalizationResult::Modified;
+                        continue;
+                    }
+                }
+            }
+
+            if !locale.id.variants.is_empty() {
+                let mut modified = Vec::new();
+                let mut unmodified = Vec::new();
+                for variant in locale.id.variants.iter() {
+                    let variant_as_tinystr: TinyStr8 = (*variant).into();
+                    if let Ok(index) = self
+                        .aliases
+                        .get()
+                        .variant
+                        .binary_search_by_key(&variant_as_tinystr, |alias| alias.0)
+                    {
+                        if let Ok(updated) = subtags::Variant::from_bytes(
+                            self.aliases.get().variant[index].1.as_bytes(),
+                        ) {
+                            modified.push(updated);
+                        }
+                    } else {
+                        unmodified.push(variant);
+                    }
+                }
+
+                if !modified.is_empty() {
+                    for variant in unmodified {
+                        modified.push(*variant);
+                    }
+                    modified.sort();
+                    modified.dedup();
+                    locale.id.variants = subtags::Variants::from_vec_unchecked(modified);
+                    result = CanonicalizationResult::Modified;
+                    continue;
+                }
+            }
+
+            // Nothing matched in this iteration, we're done.
+            break;
+        }
+
+        // Handle Locale extensions in their own loops, because these rules do not interact
+        // with each other.
+        if let Some(lang) = &locale.extensions.transform.lang {
+            let mut tlang: Locale = lang.clone().into();
+            let mut matched = false;
+            loop {
+                if uts35_check_language_rules(&mut tlang, &self.aliases)
+                    == CanonicalizationResult::Modified
+                {
+                    result = CanonicalizationResult::Modified;
+                    matched = true;
+                    continue;
+                }
+
+                break;
+            }
+
+            if matched {
+                locale.extensions.transform.lang = Some(tlang.id);
+            }
+        }
+
+        ["rg", "sd"]
+            .iter()
+            .filter_map(|key| key.parse::<Key>().ok())
```

Comment:
`suggestion`: could you avoid having to parse it every time by constructing `tinystr!` macro for `rg` and `sd` and then `Key` unchecked out of it, or even add a macro for `key!` to parse at build time?

Reply:
Also, please, add a comment explaining `rg` and `sd` here.